projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/asm/LLVMAMD64DivNode.java

```diff
@@ -0,0 +1,137 @@
+        @Child private LLVMAMD64WriteTupelNode out;
+
+        public LLVMAMD64DivwNode(LLVMAMD64WriteTupelNode out) {
+            this.out = out;
+        }
+
+        @Specialization
+        protected Object execute(VirtualFrame frame, short high, short left, short right) {
+            if (right == 0) {
+                CompilerDirectives.transferToInterpreter();
+                throw new ArithmeticException(DIV_BY_ZERO);
+            }
+            int value = Short.toUnsignedInt(high) << LLVMExpressionNode.I16_SIZE_IN_BITS | Short.toUnsignedInt(left);
+            int quotient = Integer.divideUnsigned(value, Short.toUnsignedInt(right));
+            int remainder = Integer.remainderUnsigned(value, Short.toUnsignedInt(right));
+            if (quotient > 0xFFFF) {
+                CompilerDirectives.transferToInterpreter();
+                throw new ArithmeticException(QUOTIENT_TOO_LARGE);
+            }
+            out.execute(frame, (short) quotient, (short) remainder);
+            return null;
+        }
+    }
+
+    @NodeChildren({@NodeChild("high"), @NodeChild("left"), @NodeChild("right")})
+    public abstract static class LLVMAMD64DivlNode extends LLVMExpressionNode {
+        @Child private LLVMAMD64WriteTupelNode out;
+
+        public LLVMAMD64DivlNode(LLVMAMD64WriteTupelNode out) {
+            this.out = out;
+        }
+
+        @Specialization
+        protected Object execute(VirtualFrame frame, int high, int left, int right) {
+            if (right == 0) {
+                CompilerDirectives.transferToInterpreter();
+                throw new ArithmeticException(DIV_BY_ZERO);
+            }
+            long value = Integer.toUnsignedLong(high) << LLVMExpressionNode.I32_SIZE_IN_BITS | Integer.toUnsignedLong(left);
+            long quotient = Long.divideUnsigned(value, Integer.toUnsignedLong(right));
+            long remainder = Long.remainderUnsigned(value, Integer.toUnsignedLong(right));
+            if (quotient > 0xFFFFFFFFL) {
+                CompilerDirectives.transferToInterpreter();
+                throw new ArithmeticException(QUOTIENT_TOO_LARGE);
+            }
+            out.execute(frame, (int) quotient, (int) remainder);
+            return null;
+        }
+    }
+
+    @NodeChildren({@NodeChild("high"), @NodeChild("left"), @NodeChild("right")})
+    public abstract static class LLVMAMD64DivqNode extends LLVMExpressionNode {
+        @Child private LLVMAMD64WriteTupelNode out;
+
+        public LLVMAMD64DivqNode(LLVMAMD64WriteTupelNode out) {
+            this.out = out;
+        }
+
+        @Specialization
+        protected Object execute(VirtualFrame frame, long high, long left, long right) {
+            if (right == 0) {
+                CompilerDirectives.transferToInterpreter();
+                throw new ArithmeticException(DIV_BY_ZERO);
+            }
+            LongDivision.Result result = LongDivision.divu128by64(high, left, right);
+            // TODO: error on quotient too large
```

Comment:
i would not silently swallow this error case.
Can you at least assert that? That could result in a bug that is very hard to find.

Reply:
If a program has such a bug, it will terminate with an exception (unless the exception is caught):
```
int main(void)
{
        __asm__ volatile("movw $0x8888, %ax\n"
                         "movw $0xAAAA, %dx\n"
                         "movw $1, %bx\n"
                         "divw %bx");
        return 0;
}
```
When executed:
```
[1]    11668 floating point exception (core dumped)  ./divtoobig
```
It is extremely unlikely that a "normal" program will ever trigger this exception. But yeah, I might just add the necessary checks.